integration-tests/src/tests/client/block_corruption.rs

```diff
@@ -99,3 +104,175 @@ fn change_shard_id_to_invalid() {
+        .real_epoch_managers(&genesis.config)
+        .nightshade_runtimes(&genesis)
+        .build();
+
+    let mut last_block = env.clients[0].chain.get_block_by_height(0).unwrap();
+
+    let mid_height = 3;
+    for h in 1..=mid_height {
+        let txs = create_tx_load(h, &last_block);
+        for tx in txs {
+            assert_eq!(env.clients[0].process_tx(tx, false, false), ProcessTxResponse::ValidTx);
+        }
+
+        let block = env.clients[0].produce_block(h).unwrap().unwrap();
+        env.process_block(0, block.clone(), Provenance::PRODUCED);
+        last_block = block;
+    }
+
+    let block_len = borsh::to_vec(&last_block).unwrap().len();
+    if corrupted_bit_idx >= block_len * 8 {
+        return Ok(anyhow::anyhow!("End"));
+    }
+
+    let h = mid_height + 1;
+
+    let txs = create_tx_load(h, &last_block);
+    for tx in txs {
+        assert_eq!(env.clients[0].process_tx(tx, false, false), ProcessTxResponse::ValidTx);
+    }
+
+    let correct_block = env.clients[0]
+        .produce_block(h)
+        .unwrap()
+        .with_context(|| format!("Failed to produce block at height {:?}", h))
+        .unwrap();
+
+    let mut block_vec = borsh::to_vec(&correct_block).unwrap();
+
+    // Technically we don't have to do first step, because of Ok("End") check.
+    // But we may change that part in the future for some reason, and I want this part to be reliable.
+
+    // get BIT index modulo BIT length of block
+    let mod_idx = corrupted_bit_idx % (block_vec.len() * 8);
+    // get index of BYTE that we are changing
+    let byte_idx = mod_idx / 8;
+    // get index of BIT in that BYTE that we are flipping
+    let bit_idx = mod_idx % 8;
+    // flip that BIT
+    block_vec[byte_idx] ^= 1 << bit_idx;
+
+    if let Ok(mut corrupt_block) = Block::try_from_slice(block_vec.as_slice()) {
+        corrupt_block.mut_header().get_mut().inner_rest.block_body_hash =
+            corrupt_block.compute_block_body_hash().unwrap();
+        corrupt_block.mut_header().resign(&InMemoryValidatorSigner::from_seed(
+            "test0".parse().unwrap(),
+            KeyType::ED25519,
+            "test0",
+        ));
+
+        if is_breaking_block_change(&correct_block, &corrupt_block) {
+            match env.clients[0].process_block_test(corrupt_block.clone().into(), Provenance::NONE)
+            {
+                Ok(_) => {
+                    return Err(anyhow::anyhow!(
+                        "Was able to process default block with {} bit switched.",
+                        corrupted_bit_idx
+                    ));
+                }
+                Err(e) => {
+                    if let Err(e) =
+                        env.clients[0].process_block_test(correct_block.into(), Provenance::NONE)
+                    {
+                        return Err(anyhow::anyhow!("Was unable to process default block after attempting to process default block with {} bit switched. {}",
```

Comment:
Ah, I see, nice idea! Please ignore my earlier question about processing correct block at the end. 

nit: I would move it outside of the match statement to keep it for single purpose only - to check that corrupted block is rejected.